### PR TITLE
feat(analytics): add record API for Analytics service provider Personalize

### PR DESCRIPTION
**Pull request by 5d**
#### Description of changes

- add `record` API for Analytics service provider Amazon Personalize.

#### Checklist

- [X] PR description included
- [X] `yarn test` passes
- [X] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/analytics/src/errors/validation.ts [hunk 1]
```diff
@@ -26,6 +27,10 @@ export const validationErrorMap: AmplifyErrorMap<AnalyticsValidationErrorCode> =
 			message: 'Missing region.',
 		},
 		[AnalyticsValidationErrorCode.InvalidFlushSize]: {
-			message: 'Invalid FlushSize, it should smaller than BufferSize',
+			message: 'Invalid FlushSize, it should be smaller than BufferSize',
+		},
+		[AnalyticsValidationErrorCode.NoTrackingId]: {
+			message:
+				'a trackingId field is required for using AmazonPersonalize provider',
```
Comment (jimblanc) [inline packages/analytics/src/errors/validation.ts:34]: Nit:

```suggestion
				'A trackingId is required to use Amazon Personalize',
```

File: packages/analytics/src/providers/personalize/apis/record.ts [hunk 2]
```diff
@@ -2,7 +2,95 @@
...
-export const record = (input: RecordInput): void => {
-	throw new Error('Not Yet Implemented');
+const logger = new ConsoleLogger('Personalize');
+
+export const record = ({
+	userId,
+	eventId,
+	eventType,
+	properties,
+}: RecordInput): void => {
```
Comment (jimblanc) [inline packages/analytics/src/providers/personalize/apis/record.ts:26]: Doesn't this disagree with the API review for Personalize? I believe the `userId` is in properties. The code below also seems to be grabbing it from the properties on line 41

File: packages/analytics/src/providers/personalize/utils/autoTrackMedia.ts [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+	element: HTMLMediaElement,
+	recordEvent: IRecordEvent
+) => {
+	let isPlaying = false;
+	const mediaProperties = (): Record<string, unknown> => ({
+		duration: element.duration,
+		eventValue: Number((element.currentTime / element.duration).toFixed(4)),
+	});
+
+	// TODO: don't we need to clear the interval here?
```
Comment (jimblanc) [inline packages/analytics/src/providers/personalize/utils/autoTrackMedia.ts:111]: Good question, can we do this when the DOM element is unloaded via the `unload` event?

File: packages/analytics/src/providers/personalize/utils/getEventBuffer.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+import { EventBuffer, groupBy, IAnalyticsClient } from '../../../utils';
+import { PersonalizeBufferConfig, PersonalizeBufferEvent } from '../types';
+import {
+	PersonalizeEventsClient,
+	PutEventsCommand,
+} from '@aws-sdk/client-personalize-events';
+
+const eventBufferMap: Record<string, EventBuffer<PersonalizeBufferEvent>> = {};
+const cachedClients: Record<string, PersonalizeEventsClient> = {};
```
Comment (jimblanc) [inline packages/analytics/src/providers/personalize/utils/getEventBuffer.ts:12]: Nit: Could you copy the doc comments from the other providers on what these maps track? (just to make it easier for future maintainers)

File: packages/analytics/src/providers/personalize/utils/autoTrackMedia.ts [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+			eventValue: Number((currentTime / duration).toFixed(4)),
+		};
+	};
+
+	const scriptElement = document.createElement('script');
+	scriptElement.type = 'text/javascript';
+	scriptElement.src = 'https://www.youtube.com/iframe_api';
+	document.body.append(scriptElement);
+
+	setInterval(() => {
```
Comment (jimblanc) [inline packages/analytics/src/providers/personalize/utils/autoTrackMedia.ts:60]: Do we need to clear this interval as well somewhere?

**Review by jimblanc — COMMENTED** [review]
Looking pretty good, some nits & questions

File: packages/analytics/src/errors/validation.ts (hunk 1 above)
Comment (5d) [reply packages/analytics/src/errors/validation.ts:34]: will update this.

File: packages/analytics/src/providers/personalize/apis/record.ts (hunk 2 above)
Comment (5d) [reply packages/analytics/src/providers/personalize/apis/record.ts:26]: Yes, I was trying to only use the `userId` inside the properties. However, when I porting the V5 Personalize provider logic, I found the `userId` can be either from [the event](https://github.com/aws-amplify/amplify-js/blob/main/packages/analytics/src/Providers/AmazonPersonalizeProvider.ts#L101) or [the properties](https://github.com/aws-amplify/amplify-js/blob/main/packages/analytics/src/Providers/AmazonPersonalizeProvider.ts#L97). The triggering logic for updating `userId` is quite different. I don't have complete understanding of the use cases, so I decided to just replicate the same logic from V5.

File: packages/analytics/src/providers/personalize/utils/autoTrackMedia.ts (hunk 3 above)
Comment (5d) [reply packages/analytics/src/providers/personalize/utils/autoTrackMedia.ts:111]: will add unload event to clear the timer.

File: packages/analytics/src/providers/personalize/utils/getEventBuffer.ts (hunk 4 above)
Comment (5d) [reply packages/analytics/src/providers/personalize/utils/getEventBuffer.ts:12]: will do that

File: packages/analytics/src/providers/personalize/utils/autoTrackMedia.ts (hunk 5 above)
Comment (5d) [reply packages/analytics/src/providers/personalize/utils/autoTrackMedia.ts:60]: will add `unload` event hanlder to clear the timer.

**Review by jimblanc — APPROVED** [review]

**Pull request merged**